internal/persistence/sql/migrations/single_table.go

```diff
@@ -0,0 +1,271 @@
+
+func tableFromNamespace(n *namespace.Namespace) string {
+	return fmt.Sprintf("keto_%0.10d_relation_tuples", n.ID)
+}
+
+func namespaceIDFromTable(t string) (nID int32, err error) {
+	_, err = fmt.Sscanf(t, "keto_%d_relation_tuples", &nID)
+	return nID, errors.WithStack(err)
+}
+
+func migrationTableFromNamespace(n *namespace.Namespace) string {
+	return fmt.Sprintf("keto_namespace_%0.10d_migrations", n.ID)
+}
+
+func namespaceTableFromContext(ctx context.Context) string {
+	n, ok := ctx.Value(namespaceCtxKey).(*namespace.Namespace)
+	if n == nil || !ok {
+		panic("namespace context key not set")
+	}
+	return tableFromNamespace(n)
+}
+
+func (r *relationTuple) toInternal() (*relationtuple.InternalRelationTuple, error) {
+	if r == nil {
+		return nil, nil
+	}
+
+	sub, err := relationtuple.SubjectFromString(r.Subject)
+	if err != nil {
+		return nil, err
+	}
+
+	return &relationtuple.InternalRelationTuple{
+		Relation:  r.Relation,
+		Object:    r.Object,
+		Namespace: r.Namespace.Name,
+		Subject:   sub,
+	}, nil
+}
+
+func (relationTuples) TableName(ctx context.Context) string {
+	return namespaceTableFromContext(ctx)
+}
+
+func (relationTuple) TableName(ctx context.Context) string {
+	return namespaceTableFromContext(ctx)
+}
+
+func NewToSingleTableMigrator(d dependencies) *toSingleTableMigrator {
+	return &toSingleTableMigrator{
+		d:       d,
+		perPage: 100,
+	}
+}
+
+func (m *toSingleTableMigrator) namespaceMigrationBox(n *namespace.Namespace) (*popx.MigrationBox, error) {
+	c, err := m.d.PopConnectionWithOpts(func(d *pop.ConnectionDetails) {
+		d.Options = map[string]string{
+			"migration_table_name": migrationTableFromNamespace(n),
+		}
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return popx.NewMigrationBox(
+		namespaceMigrations,
+		popx.NewMigrator(c, m.d.Logger(), nil, 0),
+		popx.WithTemplateValues(map[string]interface{}{
+			"tableName": tableFromNamespace(n),
+		}),
+	)
+}
+
+func (m *toSingleTableMigrator) getOldRelationTuples(ctx context.Context, n *namespace.Namespace, page, perPage int) (relationTuples, bool, error) {
+	q := m.d.Persister().Connection(ctx).
+		WithContext(context.WithValue(ctx, namespaceCtxKey, n)).
+		Order("object, relation, subject, commit_time").
+		Paginate(page, perPage)
+
+	var res relationTuples
+	if err := q.All(&res); err != nil {
+		return nil, false, sqlcon.HandleError(err)
+	}
+	for _, r := range res {
+		r.Namespace = n
+	}
+	return res, q.Paginator.Page < q.Paginator.TotalPages, nil
+}
+
+func (m *toSingleTableMigrator) insertOldRelationTuples(ctx context.Context, n *namespace.Namespace, rs ...*relationtuple.InternalRelationTuple) error {
+	for _, r := range rs {
+		if r.Subject == nil {
+			return errors.New("subject is not allowed to be nil")
+		}
+
+		m.d.Logger().WithFields(r.ToLoggerFields()).Trace("creating in legacy database")
+
+		if err := m.d.Persister().Connection(context.WithValue(ctx, namespaceCtxKey, n)).Create(&relationTuple{
+			ID:         "testing only",
```

Comment:
What implications does this have?

Reply:
This function is only used for testing, because we first have to insert some tuples before we can migrate them over.
`ID` is the `shard_id` column.